File: app/agents/workflow_engine.py
"""
app/agents/workflow_engine.py — 에이전트 실행 순서 및 재시도 제어
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
[최적화 — 단계별 즉시 DB 저장]
  각 에이전트 실행 완료 직후 workflow_repo.update_workflow()를 호출해
  프론트가 폴링 시 단계별 진행 상태를 즉시 확인할 수 있다.
  workflow_repo / workflow_id가 None이면 중간 저장을 건너뛴다(하위 호환).
"""

from typing import Optional

from app.agents.analyzer_agent  import AnalyzerAgent
from app.agents.matcher_agent   import MatcherAgent
from app.agents.notifier_agent  import NotifierAgent
from app.agents.planner_agent   import PlannerAgent
from app.agents.validator_agent import ValidatorAgent
from app.models.workflow import AgentContext, WorkflowStatus

MAX_RETRIES = 2
STEP_ORDER  = ["analyzer", "planner", "matcher", "validator"]


class WorkflowEngine:

    def __init__(self) -> None:
        self._agents = {
            "analyzer":  AnalyzerAgent(),
            "planner":   PlannerAgent(),
            "matcher":   MatcherAgent(),
            "validator": ValidatorAgent(),
            "notifier":  NotifierAgent(),
        }

    async def execute(
        self,
        context: AgentContext,
        *,
        workflow_repo=None,
        workflow_id: Optional[str] = None,
    ) -> AgentContext:
        """
        에이전트 파이프라인을 실행한다.
        workflow_repo + workflow_id가 주어지면 각 단계 완료 시 DB를 즉시 갱신한다.
        """
        retry_count = 0

        async def _save(ctx: AgentContext) -> None:
            """단계 완료 직후 DB 갱신 (repo가 없으면 무시)."""
            if workflow_repo and workflow_id:
                await workflow_repo.update_workflow(workflow_id, ctx.status, ctx)

        async def run_from(start_from: str | None = None) -> AgentContext:
            nonlocal retry_count, context

            start_idx = STEP_ORDER.index(start_from) if start_from else 0

            for step in STEP_ORDER[start_idx:]:
                print(f"[WorkflowEngine] running: {step}")
                context = await self._agents[step].execute(context)

                # 단계 완료 → 즉시 DB 저장 (프론트 폴링에 반영)
                await _save(context)

                if not self._agents[step].validate(context):
                    context.status = WorkflowStatus.failed
                    await _save(context)
                    return context

            # Validator 재시도 판단
            validated = context.steps.validated
            if (
                validated
                and not validated.valid
                and validated.retryFromAgent
                and validated.retryFromAgent in STEP_ORDER
                and retry_count < MAX_RETRIES
            ):
                retry_count += 1
                print(f"[WorkflowEngine] retry #{retry_count} from {validated.retryFromAgent}")
                return await run_from(validated.retryFromAgent)

            # Notifier (항상 실행, 실패해도 완료로 처리)
            context = await self._agents["notifier"].execute(context)
            context.status = WorkflowStatus.completed
            return context

        return await run_from()

File: src/agents/orchestrator/WorkflowEngine.ts
/**
 * agents/orchestrator/WorkflowEngine.ts
 * 에이전트 실행 순서를 정의하고 단계별로 실행을 위임한다.
 * 각 단계의 성공/실패를 판단하고, Validator가 재시도를 요청하면 해당 지점부터 재실행한다.
 */

import { AgentContext } from "../../types/workflow.types";
import { AgentName } from "../../types/agent.types";
import { AnalyzerAgent } from "../analyzer/AnalyzerAgent";
import { PlannerAgent } from "../planner/PlannerAgent";
import { MatcherAgent } from "../matcher/MatcherAgent";
import { ValidatorAgent } from "../validator/ValidatorAgent";
import { NotifierAgent } from "../notifier/NotifierAgent";

const MAX_RETRIES = 2;

export class WorkflowEngine {
  private analyzer = new AnalyzerAgent();
  private planner = new PlannerAgent();
  private matcher = new MatcherAgent();
  private validator = new ValidatorAgent();
  private notifier = new NotifierAgent();

  /**
   * 컨텍스트를 받아 전체 워크플로우 단계를 순서대로 실행한다.
   * 실행 순서: analyze → plan → match → validate → notify
   * Validator가 재시도를 요청하면 해당 에이전트부터 최대 MAX_RETRIES회 재실행한다.
   */
  async execute(context: AgentContext): Promise<AgentContext> {
    let retryCount = 0;
    let retryFrom: AgentName | undefined;

    const runFrom = async (startFrom?: AgentName): Promise<AgentContext> => {
      const steps: AgentName[] = ["analyzer", "planner", "matcher", "validator"];
      const startIdx = startFrom ? steps.indexOf(startFrom) : 0;

      for (let i = startIdx; i < steps.length; i++) {
        const step = steps[i];
        context = await this.runAgent(step, context);
        if (!this.validateStep(step, context)) {
          context.status = "failed";
          return context;
        }
      }

      // Validator 결과 확인
      const validated = context.steps.validated!;
      if (!validated.valid && validated.retryFromAgent && retryCount < MAX_RETRIES) {
        retryCount++;
        console.log(`[WorkflowEngine] retry #${retryCount} from ${validated.retryFromAgent}`);
        retryFrom = validated.retryFromAgent as AgentName;
        return runFrom(retryFrom);
      }

      // 알림
      context = await this.notifier.execute(context);
      context.status = "completed";
      context.updatedAt = new Date().toISOString();
      return context;
    };

    return runFrom();
  }

  private async runAgent(name: AgentName, ctx: AgentContext): Promise<AgentContext> {
    console.log(`[WorkflowEngine] running agent: ${name}`);
    switch (name) {
      case "analyzer":  return this.analyzer.execute(ctx);
      case "planner":   return this.planner.execute(ctx);
      case "matcher":   return this.matcher.execute(ctx);
      case "validator": return this.validator.execute(ctx);
      case "notifier":  return this.notifier.execute(ctx);
      default: return ctx;
    }
  }

  private validateStep(name: AgentName, ctx: AgentContext): boolean {
    switch (name) {
      case "analyzer":  return this.analyzer.validate(ctx);
      case "planner":   return this.planner.validate(ctx);
      case "matcher":   return this.matcher.validate(ctx);
      case "validator": return this.validator.validate(ctx);
      case "notifier":  return this.notifier.validate(ctx);
      default: return true;
    }
  }
}
